### PR TITLE
Fix amplitudeencoding

### PR DESCRIPTION
**Pull request by mariaschuld**
Fix for #365. 

Exchanges np.linalg.norm with hand-coded normalization. Improves and adds tests.

### CODE REVIEW

File: pennylane/templates/embeddings.py [hunk 1]
```diff
@@ -18,6 +18,7 @@
 from collections.abc import Iterable
 from pennylane.ops import RX, RY, RZ, BasisState, Squeezing, Displacement, QubitStateVector
 import numpy as np
+from pennylane.variable import Variable
```
Comment (mariaschuld) [inline pennylane/templates/embeddings.py]:
```suggestion
```
That's not needed actually...

File: pennylane/templates/embeddings.py [hunk 2]
```diff
@@ -93,28 +93,41 @@ def AmplitudeEmbedding(features, wires, pad=False, normalize=False):
...
+    if not pad and n_amplitudes != n_features:
+        raise ValueError("AmplitudeEmbedding must get a feature vector of size 2**len(wires), "
+                         "which is {}; got {}. Use ``pad=True`` to automatically pad the "
+                         "features with zeros.".format(n_amplitudes, n_features))
 
-    if normalize and np.linalg.norm(features, 2) != 1:
-        features = features * (1/np.linalg.norm(features, 2))
+    # Get normalization
+    norm = 0
+    for f in features:
```
Comment (co9olguy) [inline pennylane/templates/embeddings.py]: I suppose this will break differentiability for any feature that is encoded as a tensorflow or pytorch tensor?

File: pennylane/templates/embeddings.py [hunk 3]
```diff
@@ -93,28 +93,41 @@ def AmplitudeEmbedding(features, wires, pad=False, normalize=False):
...
+        if type(f) is Variable:
```
Comment (co9olguy) [inline pennylane/templates/embeddings.py]: Note this line is currently causing tests to fail

File: tests/test_templates_embeddings.py [hunk 4]
```diff
@@ -186,24 +186,29 @@ def circuit(x=None):
...
+        dev = qml.device('default.qubit', wires=2)
 
         @qml.qnode(dev)
         def circuit(x=None):
             AmplitudeEmbedding(features=x, wires=range(n_qubits), pad=False, normalize=False)
             return [qml.expval(qml.PauliZ(i)) for i in range(n_qubits)]
 
-        res = circuit(x=features)
-        assert np.allclose(res, [1, -1])
+        circuit(x=inpt)
```
Comment (co9olguy) [inline tests/test_templates_embeddings.py]: How come you pass as a keyword arg? Does it not work if the input is a regular argument?

File: tests/test_templates_embeddings.py [hunk 5]
```diff
@@ -215,26 +220,68 @@ def circuit(x=None):
 
         with pytest.raises(ValueError) as excinfo:
             circuit(x=[np.sqrt(0.2), np.sqrt(0.8), 0, 0, 0])
-        assert excinfo.value.args[0] == "AmplitudeEmbedding requires the size of feature vector to be smaller than or equal to 2**len(wires), which is 4; got 5."
+        assert excinfo.value.args[0] == "AmplitudeEmbedding requires the size of feature vector to " \
```
Comment (co9olguy) [inline tests/test_templates_embeddings.py]: No biggie, but for future reference, you can also use:
```
with pytest.raises(ValueError, match="some string to match")
```
This allows you to just match a substring, rather than having to type the whole string out here in tests

File: tests/test_templates_embeddings.py [hunk 6]
```diff
@@ -215,26 +220,68 @@ def circuit(x=None):
...
+        argument if `normalize` is set to `True`."""
+
+        dev = qml.device('default.qubit', wires=1)
+
+        @qml.qnode(dev)
+        def circuit(x):
+            AmplitudeEmbedding(features=x, wires=0, pad=False, normalize=True)
+            return qml.expval(qml.PauliZ(0))
+
+        circuit(inpt)
```
Comment (co9olguy) [inline tests/test_templates_embeddings.py]: ok I see here that you can pass the input as a positional argument. How come the other tests use the keyword-arg approach?

File: pennylane/templates/embeddings.py (hunk 2 above)
Comment (mariaschuld) [reply pennylane/templates/embeddings.py]: Ah, shoot. You are right. But how else would we do it? Check for the type (which is very restrictive)? Note that this is a general problem in templates, how can we interact with inputs without assuming too much.

File: tests/test_templates_embeddings.py (hunk 5 above)
Comment (mariaschuld) [reply tests/test_templates_embeddings.py]: We had an issue with the raises() function before in a similar example...but yes, else that would be more elegant.

File: pennylane/templates/embeddings.py (hunk 2 above)
Comment (josh146) [reply pennylane/templates/embeddings.py]: A more drastic option is to remove normalization altogether, and leave it up to the plugins to check. Previously, the problem was that _nothing_ was checking normalization, but this has been slowly added to all plugins (#346 adds this to default.qubit, not yet merged).

This way the user still gets a useful error message, and the checking occurs at a point where autodiff 'doesn't matter' per se.

File: tests/test_templates_embeddings.py (hunk 6 above)
Comment (mariaschuld) [reply tests/test_templates_embeddings.py]: See above.

File: tests/test_templates_embeddings.py (hunk 4 above)
Comment (mariaschuld) [reply tests/test_templates_embeddings.py]: This is the test for a non-variable (i.e., a numpy array), since it gets passed via keyword argument. Below is the test for a list of Variable objects, which uses positional arguments.

File: pennylane/templates/embeddings.py (hunk 3 above)
Comment (mariaschuld) [reply pennylane/templates/embeddings.py]: I see...only ran the templates tests, but configuration fails.

I need to rethink this. We had trouble with in-template-checks before. Do you have an idea how to cater for arrays, lists, variables, tensors etc at the same time? 

I start thinking that we just do not use any checks in templates, since it is strictly speaking inside a quantum function that should only do quantum operations...

File: pennylane/templates/embeddings.py (hunk 2 above)
Comment (co9olguy) [reply pennylane/templates/embeddings.py]: Some thoughts on how to deal with this:

- we have a dictionary which maps different arithmetic operations to the equivalent in the relevant library (e.g., https://www.tensorflow.org/api_docs/python/tf/math/l2_normalize). This works, but is clunky (and requires checking if libraries are installed via `try/except`.

- since all those libraries should overload the basic arithmetic operations, we could do it implicitly, e.g., `sum(array ** 2)`. If `array` is a list (or numpy array) of `Tensor`s, this should work. Otherwise if array itself is a `Tensor`, you would have to manually do the sum: `sum([b[idx] for idx in range(b.shape[0])])`.

Comment (josh146) [reply pennylane/templates/embeddings.py]: > I suppose this will break differentiability for any feature that is encoded as a tensorflow or pytorch tensor?

Actually just realised that nothing should be a Tensor at this point in the code. The QNode converts from Tensor -> NumPy array -> Variable (if differentiable).

File: pennylane/templates/embeddings.py (hunk 3 above)
Comment (josh146) [reply pennylane/templates/embeddings.py]: > I start thinking that we just do not use any checks in templates, since it is strictly speaking inside a quantum function that should only do quantum operations...
I agree :+1:

File: pennylane/templates/embeddings.py (hunk 2 above)
Comment (co9olguy) [reply pennylane/templates/embeddings.py]: Then we just need to make sure that `+` and `**` are overloaded for pennylane `Variable`s

Comment (mariaschuld) [reply pennylane/templates/embeddings.py]: Actually, the first thing to do is to write integration tests for the embeddings, to check if the interfaces pass through. 
I only realize this now, the templates were supposed to have had them all along.

Comment (mariaschuld) [reply pennylane/templates/embeddings.py]: I tested the torch and tf interface, and as @josh146 said everything works fine with the new manual normalization, the operators must be already overloaded.

File: pennylane/templates/embeddings.py (hunk 3 above)
Comment (mariaschuld) [reply pennylane/templates/embeddings.py]: For now I will leave as is, and we can think about a general strategy elsewhere.

Comment (mariaschuld) [reply pennylane/templates/embeddings.py]: The test passes now, except from a version error in the about() function that is also in master.

**Comment by mariaschuld** [comment]
I addressed all comments and the fix is working with all interfaces and argument passing strategies. Ready for re-review.

**Comment by co9olguy** [comment]
closes #365
